libnest/src/package/npf.rs

```diff
@@ -26,40 +31,83 @@ impl<'explorer> NPFFile<'explorer> {
 /// Structure representing an NPF to allow interacting with it
 #[derive(Debug)]
 pub struct NPFExplorer {
+    manifest: Manifest,
     path: PathBuf,
 }
 
 impl NPFExplorer {
+    fn load_manifest(path: &Path) -> Result<Manifest, NPFExplorationError> {
+        let mut file = File::open(path.join("manifest.toml")).map_err(|err| match err.kind() {
+            std::io::ErrorKind::NotFound => NPFExplorationErrorKind::MissingManifest,
+            _ => NPFExplorationErrorKind::FileOpenError(path.to_path_buf()),
+        })?;
+
+        let mut content = String::new();
+        file.read_to_string(&mut content)
+            .map_err(|_| FileOpenError(path.to_path_buf()))?;
+
+        Ok(toml::from_str(&content).map_err(|_| NPFExplorationErrorKind::InvalidManifest)?)
+    }
+
     /// Create a NPFExplorer from a name and a path to an NPF archive
-    pub fn from(name: &str, npf_path: &Path) -> Result<Self, Error> {
+    pub fn from(name: &str, npf_path: &Path) -> Result<Self, NPFExplorationError> {
         let path = std::env::temp_dir().join(name);
 
         // Create a directory to extract the NPF
-        fs::create_dir(&path)?;
+        fs::create_dir(&path).map_err(|_| NPFExplorationErrorKind::UnpackError)?;
 
         // Unpack the NPF
-        let file = File::open(npf_path)?;
-        let mut archive = Archive::new(&file);
-        archive.unpack(&path)?;
+        File::open(npf_path)
+            .and_then(|file| {
+                let mut archive = Archive::new(&file);
+                archive.unpack(&path)
+            })
+            .map_err(|_| NPFExplorationErrorKind::UnpackError)?;
+
+        let manifest = Self::load_manifest(&path)?;
 
-        Ok(Self { path })
+        Ok(Self { path, manifest })
     }
 
     /// Retrieves a handle over a file in the NPF
-    fn get_file<P: AsRef<Path>>(&self, path: P) -> Result<NPFFile, Error> {
-        let file = File::open(self.path.join(path))?;
+    fn get_file(&self, path: &Path) -> Result<NPFFile, NPFExplorationError> {
+        let file = File::open(self.path.join(path)).map_err(|err| match err.kind() {
+            std::io::ErrorKind::NotFound => {
+                NPFExplorationErrorKind::FileNotFound(path.to_path_buf())
+            }
+            _ => NPFExplorationErrorKind::FileOpenError(path.to_path_buf()),
+        })?;
 
         Ok(NPFFile::from(file, PhantomData))
     }
 
-    /// Retrieves a handle over the NPF's manifest.toml
-    pub fn get_manifest(&self) -> Result<NPFFile, Error> {
-        self.get_file("manifest.toml")
+    /// Retrieves the NPF's manifest
+    pub fn manifest(&self) -> &Manifest {
+        &self.manifest
     }
 
     /// Retrieves a handle over the NPF's data.tar.gz
-    pub fn get_data(&self) -> Result<NPFFile, Error> {
-        self.get_file("data.tar.gz")
+    pub fn get_data(&self) -> Result<Option<NPFFile>, NPFExplorationError> {
```

Comment:
The name `get_data` is anticonventional in Rust, could you elaborate for such name and not a more common one, like `files()` or `data()`?

Reply:
Usually, I pick `get_xxx` for things that require an action to be performed in order to succeed, and only `xxx` for purely data-based getters.
However, I trust your Rust-wisdom on this, please correct me if I'm begin anti-idiomatic.

Reply:
I think the best move is to put the action within the name of the function, like `open_data()` or anything that implies an action that takes time and isn't straightforward.

For a lot of OO programmers, `get_xxx()` is a getter, and even though Rust isn't an OO language, we should avoid facing the ones with OO experience like that.

Reply:
Makes sense, `open_data` seems good to me